### PR TITLE
Add graph manipulation and utilities API

### LINKED ISSUE #117: Introduce public graph manipulation API
I have been reading a bit into [torch.fx](https://pytorch.org/docs/stable/fx.html) and how that makes it possible to implement custom compiler passes and other cool graph manipulations in PyTorch. I would like to expose a somewhat similar API to Axon to allow for easy manipulation of the Axon graph.

The main Axon API is basically a DSL for creating and composing graphs. These graphs (or Axons as I like to call them) have the following (recursive) typedef:

```elixir
@type axon :: Axon.t() | tuple(axon())
```

The axon  data structure is a tree-like data structure. I am proposing to add tree manipulation functions which look something like:

```elixir
tree_lower(graph :: axon(), any(), (axon(), any() -> {any(), axon()})) :: any() # Unsure of name

# Conveniences around the tree_lower
tree_reduce(graph :: axon(), any(), (axon(), any() -> any()) :: any() 
tree_map(graph :: axon(), (axon -> axon)) :: axon()
```

`tree_lower` takes in an entire graph, an initial accumulator, and a function which takes in the current node and the accumulator and returns the updated accumulator and the next node to traverse. Returning the next node can be useful specifically for writing compiler passes in which certain patterns might be fused or changed in some way. The `lower` comes from the fact that this function can be used as the fundamental function for lowering from Axon to Nx or another representation. The Axon compiler will be rewritten in this way, and so will Axon ONNX.

Both `tree_reduce` and `tree_map` are conveniences around the lowering operation that don't have to concern themselves with traversal (how they're traversed is pretty much set in place). `tree_map` will be used to manipulate node manipulations such as parameter freezing and mixed precision. `tree_reduce` will be used for graph convenience methods like `get_parameters` and `get_inputs`.

I think this will kind of help unify the implementations of AxonOnnx and the Axon Nx Expr compiler. I am considering other abstractions around the Axon compiler, but this is the one I wanted to expose first.

### PR DESCRIPTION
**Pull request by seanmor5**
Resolves #117 

I added a bunch of additional metadata utilities, since there was no public API for that yet and exposing an API makes it more clear how to manipulate parts of the Axon struct. Additionally, it allows us to validate internally to avoid messed up graphs and things going wrong at execution time

### CODE REVIEW

File: lib/axon.ex [hunk 1]
```diff
@@ -3139,93 +3139,327 @@ defmodule Axon do
...
+    %Axon{output_shape: parent_shape} = parent
+    # TODO: compatible? does not handle container shapes
+    unless Axon.Shape.compatible?(parent_shape, shape) do
+      raise ArgumentError,
+            "new parent graph must be compatible with old parent" <>
+              " graph, but got incompatible shapes #{inspect(parent_shape)}" <>
+              " and #{inspect(shape)}"
+    end
 
-  def tree_reduce(%Axon{parent: x} = axon, acc, fun) when is_list(x) do
```
Comment (seanmor5) [inline lib/axon.ex:3164]: I guess I should probably deprecate this instead?

Comment (josevalim) [reply lib/axon.ex:3164]: Nah, it is fine. :) It is very early, best to move forward.

**Comment by josevalim** [comment]
Looks great! My only suggestion would be to tag those functions as `@doc type: :traversal` perhaps? Or alternatively move them to an `Axon.Graph` module or similar, in case `Axon` is growing too large.

**Comment by seanmor5** [comment]
@josevalim I may choose to reorganize the API so some functions are in different modules before the next release, but for now I am okay with leaving them at the top-level :)

**Pull request merged**